### PR TITLE
Issue25329 fix cors issue when no internet

### LINKED ISSUE #25329: [Task]: CORS issue shown when trying run example without internet connection
### What needs to happen?

CORS issue shown when trying run example without internet connection

### Issue Priority

Priority: 3 (nice-to-have improvement)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [X] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### LINKED ISSUE #25331: [Bug]: I can cancel code running without internet connection
### What happened?

Steps to reproduce:

1. Change default example code (changing comment is enough)
2. Run it
3. Turn off internet connection
4. Cancel example

Expected result: Error canceling running code, because internet connection unavailable
Actual result: Running code cancelled

### Issue Priority

Priority: 3 (minor)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [X] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### LINKED ISSUE #25332: [Bug]: can't scroll tags in example selector on macOS using mouse
### What happened?

Steps to reproduce (macOS):

1. Open example selector
2. Try to scroll tags using mouse

Actual result: tags don't scrolled

### Issue Priority

Priority: 3 (minor)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### LINKED ISSUE #25333: [Task]: Displaying tags conveniently
### What needs to happen?

Sort tags to start on selecting or show all tags withiout scrolling in example selector

### Issue Priority

Priority: 3 (nice-to-have improvement)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [X] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### LINKED ISSUE #25336: [Bug]: Possible to run 2 examples simultaneously
### What happened?

Steps to reproduce:

1. Change code of example
2. Run it
3. Switch to another example
4. Change code
5. Run it

Expected result: First example should be canceled when switching to the second
Actual result: Fist example don't cancelled

### Issue Priority

Priority: 3 (minor)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [x] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by Malarg**
resolves #25329
resolves #25331
resolves #25336
resolves #25332
resolves #25333

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Malarg** [comment]
@alexeyinkin @nausharipov fyi

File: playground/frontend/playground_components/lib/src/util/connectivity_result.dart [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+import 'package:connectivity_plus/connectivity_plus.dart';
+
+extension ConnectivityResultExtension on ConnectivityResult {
+  bool get isConnected => [
+        ConnectivityResult.ethernet,
+        ConnectivityResult.mobile,
+        ConnectivityResult.vpn,
+        ConnectivityResult.wifi,
+      ].contains(this);
```
Comment (alexeyinkin) [inline playground/frontend/playground_components/lib/src/util/connectivity_result.dart:27]: Check if != none.

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: playground/frontend/playground_components/lib/src/controllers/playground_controller.dart [hunk 2]
```diff
@@ -49,6 +49,7 @@ const kTitle = 'Catalog';
 const kExecutionCancelledText = '\nPipeline cancelled';
 const kPipelineOptionsParseError =
     'Failed to parse pipeline options, please check the format (example: --key1 value1 --key2 value2), only alphanumeric and ",*,/,-,:,;,\',. symbols are allowed';
+const kInternetConnectionUnavailable = 'Internet connection unavailable';
 const kCachedResultsLog =
     'The results of this example are taken from the Apache Beam Playground cache.\n';
```
[the remark this answers is not in the document]
Comment (Malarg) [reply playground/frontend/playground_components/lib/src/controllers/playground_controller.dart]: done

File: playground/frontend/playground_components/lib/src/repositories/code_client/grpc_code_client.dart [hunk 3]
```diff
@@ -18,10 +18,9 @@
 
 import 'package:grpc/grpc.dart';
 
+import '../../../playground_components.dart';
```
[the remark this answers is not in the document]
Comment (Malarg) [reply playground/frontend/playground_components/lib/src/repositories/code_client/grpc_code_client.dart]: done

File: playground/frontend/playground_components/lib/src/repositories/code_client/grpc_code_client.dart [hunk 4]
```diff
@@ -192,6 +194,11 @@ class GrpcCodeClient implements CodeClient {
     try {
       return await invoke();
     } on GrpcError catch (error) {
+      //TODO(Malarg): check if internet connection available here using connectivity_plus
+      //throw internet unavailable exception, if not
+      if (error.message?.contains('CORS') ?? false) {
+        throw const RunCodeError(message: kInternetConnectionUnavailable);
+      }
```
[the remark this answers is not in the document]
Comment (Malarg) [reply playground/frontend/playground_components/lib/src/repositories/code_client/grpc_code_client.dart]: Bug gone from chrome, changed error text

File: playground/frontend/playground_components/lib/src/util/connectivity_result.dart (hunk 1 above)
Comment (Malarg) [reply playground/frontend/playground_components/lib/src/util/connectivity_result.dart]: done

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: playground/frontend/playground_components/assets/translations/en.yaml [hunk 5]
```diff
@@ -17,12 +17,16 @@
 
 errors:
   error: 'Error'
+  failedParseOptions: "Failed to parse pipeline options, please check the format (example: --key1 value1 --key2 value2), only alphanumeric and \",*,/,-,:,;,',. symbols are allowed"
+  internetUnavailable: 'Internet connection is unavailable'
   loading: 'Error while loading.'
   loadingCatalog: 'Cannot load the example catalog.'
   loadingExample: 'Cannot load the example.'
   savingSnippet: 'Cannot create a shareable link.'
+  unknownError: 'Unknown error. Potential internet connection problem or CORS issue'
```
Comment (Malarg) [inline playground/frontend/playground_components/assets/translations/en.yaml:26]: possibly

File: playground/frontend/playground_components/assets/translations/en.yaml [hunk 6]
```diff
@@ -17,12 +17,16 @@
 
 errors:
   error: 'Error'
+  failedParseOptions: "Failed to parse pipeline options, please check the format (example: --key1 value1 --key2 value2), only alphanumeric and \",*,/,-,:,;,',. symbols are allowed"
```
[the remark this answers is not in the document]
Comment (Malarg) [reply playground/frontend/playground_components/assets/translations/en.yaml:20]: No, it's necessary for pasting single quote symbol

File: playground/frontend/playground_components/assets/translations/en.yaml (hunk 5 above)
Comment (Malarg) [reply playground/frontend/playground_components/assets/translations/en.yaml]: done

File: playground/frontend/playground_components/lib/src/repositories/models/run_code_result.dart [hunk 7]
```diff
@@ -33,13 +33,13 @@ enum RunCodeStatus {
...
-  RunCodeStatus.timeout,
   RunCodeStatus.compileError,
+  RunCodeStatus.finished,
+  RunCodeStatus.preparationError,
   RunCodeStatus.runError,
+  RunCodeStatus.timeout,
+  RunCodeStatus.unknownError,
   RunCodeStatus.validationError,
-  RunCodeStatus.preparationError,
-  RunCodeStatus.finished,
```
[the remark this answers is not in the document]
Comment (Malarg) [reply playground/frontend/playground_components/lib/src/repositories/models/run_code_result.dart]: done

File: playground/frontend/playground_components/lib/src/controllers/code_runner.dart [hunk 8]
```diff
@@ -122,8 +135,21 @@ class CodeRunner extends ChangeNotifier {
...
+        errorMessage: 'errors.internetUnavailable'.tr(),
+        graph: _result?.graph,
+      );
+      notifyListeners();
+      return;
+    }
+
     snippetEditingController = null;
-    await _runSubscription?.cancel();
+    unawaited(_runSubscription?.cancel());
```
[the remark this answers is not in the document]
Comment (Malarg) [reply playground/frontend/playground_components/lib/src/controllers/code_runner.dart]: done

File: playground/frontend/lib/modules/examples/components/filter/filter.dart [hunk 9]
```diff
@@ -77,22 +78,29 @@ class _Types extends StatelessWidget {
 }
 
 class _Tags extends StatelessWidget {
-  const _Tags();
+  final ScrollController scrollController = ScrollController();
```
[the remark this answers is not in the document]
Comment (Malarg) [reply playground/frontend/lib/modules/examples/components/filter/filter.dart]: done

File: playground/frontend/lib/pages/standalone_playground/notifiers/example_selector_state.dart [hunk 10]
```diff
@@ -47,14 +47,28 @@ class ExampleSelectorState with ChangeNotifier {
...
   }
 
+  void _sortTagsBySelected() {
+    tags.sort((a, b) {
+      if (selectedTags.contains(a) && !selectedTags.contains(b)) {
+        return -1;
+      } else if (!selectedTags.contains(a) && selectedTags.contains(b)) {
+        return 1;
+      } else {
+        return 0;
```
[the remark this answers is not in the document]
Comment (Malarg) [reply playground/frontend/lib/pages/standalone_playground/notifiers/example_selector_state.dart]: done

File: playground/frontend/playground_components/lib/src/repositories/code_client/grpc_code_client.dart [hunk 11]
```diff
@@ -192,6 +196,9 @@ class GrpcCodeClient implements CodeClient {
     try {
       return await invoke();
     } on GrpcError catch (error) {
+      if (error.code == StatusCode.unknown) {
+        throw RunCodeError(message: 'errors.unknownError'.tr());
+      }
```
Comment (alexeyinkin) [inline playground/frontend/playground_components/lib/src/repositories/code_client/grpc_code_client.dart:201]: Explain why this is needed.

File: playground/frontend/lib/pages/standalone_playground/screen.dart [hunk 12]
```diff
@@ -37,6 +37,7 @@ import 'widgets/playground_page_footer.dart';
 import 'widgets/playground_page_providers.dart';
 
 class StandalonePlaygroundScreen extends StatelessWidget {
+  static const kAppBarButtonsWidth = 1105;
```
Comment (alexeyinkin) [inline playground/frontend/lib/pages/standalone_playground/screen.dart:40]: Explain how this was obtained. Add a TODO and link to an issue to do some automatic calculation.

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: playground/frontend/lib/pages/standalone_playground/screen.dart (hunk 12 above)
Comment (Malarg) [reply playground/frontend/lib/pages/standalone_playground/screen.dart:44]: done

File: playground/frontend/playground_components/lib/src/repositories/code_client/grpc_code_client.dart (hunk 11 above)
Comment (Malarg) [reply playground/frontend/playground_components/lib/src/repositories/code_client/grpc_code_client.dart:203]: done

File: playground/frontend/playground_components/lib/src/controllers/code_runner.dart [hunk 13]
```diff
@@ -122,8 +135,23 @@ class CodeRunner extends ChangeNotifier {
...
+        errorMessage: 'errors.internetUnavailable'.tr(),
+        graph: _result?.graph,
+      );
+      notifyListeners();
+      return;
+    }
+
     snippetEditingController = null;
-    await _runSubscription?.cancel();
+    // Awaited cancelling subscription here blocks futrher method execution.
```
[the remark this answers is not in the document]
Comment (Malarg) [reply playground/frontend/playground_components/lib/src/controllers/code_runner.dart]: fixed

File: playground/frontend/lib/pages/standalone_playground/notifiers/example_selector_state.dart [hunk 14]
```diff
@@ -47,28 +49,61 @@ class ExampleSelectorState with ChangeNotifier {
...
     notifyListeners();
   }
 
   void removeSelectedTag(String tag) {
     selectedTags.remove(tag);
+    _sortTagsBySelected();
     notifyListeners();
   }
 
+  void _sortTagsBySelected() {
```
[the remark this answers is not in the document]
Comment (Malarg) [reply playground/frontend/lib/pages/standalone_playground/notifiers/example_selector_state.dart]: done

File: playground/frontend/lib/pages/standalone_playground/notifiers/example_selector_state.dart [hunk 15]
```diff
@@ -47,28 +49,61 @@ class ExampleSelectorState with ChangeNotifier {
...
+            ? entry2.value.compareTo(entry1.value)
+            : entry2.key.compareTo(entry1.key),
+      );
+    return tagEntries.map((entry) => entry.key).toList();
+  }
+
+  Map<String, int> _buildTagsFrequencyMap(
+    List<CategoryWithExamples> categories,
+  ) {
+    Map<String, int> result = {};
```
[the remark this answers is not in the document]
Comment (Malarg) [reply playground/frontend/lib/pages/standalone_playground/notifiers/example_selector_state.dart]: done

**Review by alexeyinkin — APPROVED** [review]

**Pull request closed without merging**